common/formatting/layout_optimizer.h

```diff
@@ -0,0 +1,445 @@
+// Copyright 2017-2021 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Improve code formatting by optimizing token partitions layout with
+// algorithm documented in https://research.google/pubs/pub44667/
+// (similar tool for R language: https://github.com/google/rfmt)
+
+#ifndef VERIBLE_VERILOG_FORMATTING_LAYOUT_OPTIMIZER_H_
+#define VERIBLE_VERILOG_FORMATTING_LAYOUT_OPTIMIZER_H_
+
+#include <deque>
+#include <vector>
+
+#include "common/formatting/basic_format_style.h"
+#include "common/formatting/token_partition_tree.h"
+#include "common/formatting/unwrapped_line.h"
+#include "common/util/vector_tree.h"
+
+namespace verible {
+
+// Layout type
+enum class LayoutType {
+  // Holds UnwrappedLine
+  kLayoutLine,
+
+  // Merges sublayouts, horizontally
+  kLayoutHorizontalMerge,
+
+  // Merges sublayout, vertically
+  kLayoutVerticalMerge,
+
+  // Indent sublayout
+  // ----------------
+  // Why do we introduce new layout instead of using
+  // kLayoutHorizontalMerge + kLayoutLine(empty + indent)?
+  // Because:
+  // 1. It does not introduce Knot at column_limit - indent
+  // 2. No need to check layouts type in HorizontalJoin() whether to
+  //    skip or not before_spaces (when merging indent line with normal
+  //    line/layout)
+  // 3. Wrapping layout forces GetSpacesBefore() return 0
+  kLayoutIndent,
+};
+
+std::ostream& operator<<(std::ostream& stream, const LayoutType& type);
+
+// Intermediate partition tree layout
+class Layout {
+ public:
+  Layout(LayoutType type, int spacing)
+      : type_(type), indentation_(0), spaces_before_(spacing) {}
+
+  Layout(const UnwrappedLine& uwline)
+      : type_(LayoutType::kLayoutLine), indentation_(0) {
+    tokens_ = uwline.TokensRange();
+    if (tokens_.size() > 0) {
```

Comment:
For container size queries to determine if there is something in there (foo.size() != 0, foo.size() >= 1, etc.), we should use `empty()`, because the empty() functionality might be faster depending on the container (for most containers that is `O(1)`, but size() might be `O(something depending on N)`)

so here, that would be `if (!tokens_.empty()) ...`

(also other places in this PR).

Reply:
Done